### DISCUSSION TITLE
Roda + Omniauth

### DISCUSSION THREAD

**Discussion by kez** [jeremyevans/roda#299] [General]
Hi all - happy new year! 

I was wondering if anyone has implemented Omniauth into a modern Roda app using the `route_csrf` plugin?  I've integrated it before but with some hoop jumping with CSRF checks that I want to avoid.

My understanding is `route_csrf` will write tokens into `_csrf` but OmniAuth is looking for the "traditional" `authenticity_token` value.  I would love to *just* use `route_csrf`.  

I believe OmniAuth will want a custom TokenVerifier class (example below).  I'm not as good with Roda internals as I'd like to be, but could this example class potentially have access to Roda's `env/request` and the plugin's `valid_csrf?` method?

Any insight appreciated!

```ruby
# as detailed https://github.com/omniauth/omniauth/releases/tag/v2.0.0
class TokenVerifier
  include ActiveSupport::Configurable
  include ActionController::RequestForgeryProtection

  def call(env)
    @request = ActionDispatch::Request.new(env.dup)
    raise OmniAuth::AuthenticityError unless verified_request?
  end

  private
  attr_reader :request
  delegate :params, :session, to: :request
end
# in an initializer
OmniAuth.config.request_validation_phase = TokenVerifier.new
```

**Comment by jeremyevans** [comment]
You should probably try https://github.com/janko/rodauth-omniauth

**Comment by janko** [comment]
If you're not using Rodauth, you should be able to just initialize the Roda app from the Rack env and call `check_csrf!` on it for the request validation phase:

```rb
OmniAuth.config.request_validation_phase = -> (env) { App.new(env).check_csrf! }
```

The mentioned rodauth-omniauth gem takes a more elaborate approach, where it runs the OmniAuth app via `r.run`, and saves the Roda instance in the Rack env hash, so that it's available within OmniAuth hooks. But it's probably not necessary for common use cases.

**Reply by jeremyevans** [reply]
This is much better advice than mine.  I though I was in a Rodauth discussion when I posted my comment, sorry.

**Reply by kez** [reply]
Thanks both.  That's a super helpful, succinct example @janko!  In theory OmniAuth's CSRF could probably be disabled  as route_csrf should be handling it, but feels better to have OmniAuth checking.
